Source/Core/Core/HW/EXI/EXI_DeviceModem.cpp

```diff
@@ -0,0 +1,396 @@
+// Copyright 2024 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Core/HW/EXI/EXI_DeviceModem.h"
+
+#include <inttypes.h>
+
+#include <algorithm>
+#include <memory>
+#include <optional>
+#include <string>
+
+#include "Common/BitUtils.h"
+#include "Common/ChunkFile.h"
+#include "Common/CommonTypes.h"
+#include "Common/Logging/Log.h"
+#include "Common/Network.h"
+#include "Common/StringUtil.h"
+#include "Core/Config/MainSettings.h"
+#include "Core/CoreTiming.h"
+#include "Core/HW/EXI/EXI.h"
+#include "Core/HW/Memmap.h"
+#include "Core/PowerPC/PowerPC.h"
+#include "Core/System.h"
+
+namespace ExpansionInterface
+{
+
+CEXIModem::CEXIModem(Core::System& system, ModemDeviceType type) : IEXIDevice(system)
+{
+  switch (type)
+  {
+  case ModemDeviceType::TAPSERVER:
+    m_network_interface = std::make_unique<TAPServerNetworkInterface>(
+        this, Config::Get(Config::MAIN_MODEM_TAPSERVER_DESTINATION));
+    INFO_LOG_FMT(SP1, "Created tapserver physical network interface.");
+    break;
+  }
+
+  m_regs[Register::DEVICE_TYPE] = 0x02;
+  m_regs[Register::INTERRUPT_MASK] = 0x02;
+}
+
+CEXIModem::~CEXIModem()
+{
+  m_network_interface->Deactivate();
+}
+
+bool CEXIModem::IsPresent() const
+{
+  return true;
+}
+
+void CEXIModem::SetCS(int cs)
+{
+  m_transfer_descriptor = INVALID_TRANSFER_DESCRIPTOR;
+}
+
+bool CEXIModem::IsInterruptSet()
+{
+  return !!(m_regs[Register::INTERRUPT_MASK] & m_regs[Register::PENDING_INTERRUPT_MASK]);
+}
+
+void CEXIModem::ImmWrite(u32 data, u32 size)
+{
+  if (m_transfer_descriptor == INVALID_TRANSFER_DESCRIPTOR)
+  {
+    m_transfer_descriptor = data;
+    if (m_transfer_descriptor == 0x00008000)
+    {  // Reset
+      m_network_interface->Deactivate();
+      m_transfer_descriptor = INVALID_TRANSFER_DESCRIPTOR;
+    }
+  }
+  else if (!IsWriteTransfer(m_transfer_descriptor))
+  {
+    ERROR_LOG_FMT(SP1, "Received EXI IMM write {:x} ({} bytes) after read command {:x}", data, size,
+                  m_transfer_descriptor);
+    m_transfer_descriptor = INVALID_TRANSFER_DESCRIPTOR;
+  }
+  else if (IsModemTransfer(m_transfer_descriptor))
+  {  // Write AT command buffer or packet send buffer
+    const u32 be_data = htonl(data);
+    HandleWriteModemTransfer(&be_data, size);
+  }
+  else
+  {  // Write device register
+    u8 reg_num = static_cast<uint8_t>((m_transfer_descriptor >> 24) & 0x1F);
+    bool should_update_interrupts = false;
+    for (; size; size--)
+    {
+      if (reg_num >= m_regs.size())
+      {
+        break;
+      }
```

Comment:
Can't this condition be part of the `for`'s condition _(e.g. `size && reg_num < m_regs.size()`)_?